### PR TITLE
AP20-813: add getObjectSize to size an object without downloading it

### PR DESCRIPTION
**Pull request by ruslan-avantis**
## What & why

Adds `getObjectSize(url)` — the byte size of an object, without downloading it.

Sizing belongs next to streaming. This package already owns the location grammar (`RX_S3`, `RX_FILE`) and the `{ $src }` unwrapping, so a size helper living anywhere else has to restate both and can only ever cover a subset of the locations `createReadStream` accepts. That is exactly what happened: cn-core grew a local `getS3ObjectSize` with a copy of the URI regex that handled neither `file://` nor range locations, and a reviewer rightly flagged the mismatch. This PR is where that helper should have gone; the cn-core copy is being deleted.

For S3 it is a `headObject` call, so only metadata crosses the wire and a caller can decide *how* to send a file before reading a byte of it.

Motivating case ([AP20-813](https://projectcx.atlassian.net/browse/AP20-813)): a document-ingest endpoint caps a single request at 25 MB and requires a three-step signed upload above that. The connector has to pick a path per file, and downloading a 27 MB PDF just to measure it is not an option inside a 128 MB lambda.

## Behaviour

Accepts everything `createReadStream` accepts, so anything streamable is also sizeable:

| Location | Result |
|---|---|
| `s3://bucket/key` | `headObject` → `ContentLength` |
| `https://s3.amazonaws.com/bucket/key` | same |
| `{ $src: 's3://…' }` | unwrapped first, same as `createReadStream` |
| `file:///path` | `fs.statSync().size` |

It answers for the stream `createReadStream` would produce, so a range location returns its `length` and makes no request at all. Ranges on `file://` are rejected, because `createReadStream` does not honour them there either — and `RX_FILE` being greedy, the suffix would otherwise land inside the path and surface as ENOENT for a file that never existed.

A missing `ContentLength` rejects: returning `0` would read as "small" to a caller routing by size, the exact mistake this helper exists to prevent. A genuinely empty object still returns `0`.

## Also fixes: `createReadStream` read one byte too many on every range

An HTTP byte range is inclusive at both ends, so the last byte of a `length`-byte slice starting at `offset` sits at `offset + length - 1`. The code asked for `offset + length`.

Confirmed against a live workflow payload (`?offset=2706&length=332`):

- 332 bytes is exactly one JSON record, ending `…"COMPANY"}]`
- the 333rd byte is the `\n` separating the next record

That is why it has gone unnoticed: the writers of these ranges separate records with a newline, and `JSON.parse` ignores trailing whitespace. Any range read whose next byte is not whitespace has been getting corrupt data.

Verified after the fix, against the same live object: `loadJson` returns the single expected record, `getObjectSize` on the range returns 332 with no request, and on the whole object 26,531.

This is what makes the range branch of `getObjectSize` honest, so both land together — shipping the size helper with a refusal whose stated reason is a bug we could fix in the same diff would leave an unexplained restriction behind.

The function is `async`, so every failure — bad argument, unsupported location, missing local file, S3 error — reaches the caller as a rejection rather than as a throw while the argument is being evaluated. `TypeError` for a missing or non-string url, `Error` for an unsupported format.

`aws-sdk` moves from `devDependencies` to `peerDependencies`. The entry point requires it at load time, so the contract needs to be declared; a plain `dependency` would add the whole SDK v2 to every consumer bundle when the Lambda runtime already provides it.

## Tests

Both new suites stub `aws.S3` and run fully offline, following the pattern of the existing `createWriteStream.options` unit test.

`tests/unit/getObjectSize.test.js` covers both accepted URI forms, the `{ $src }` envelope, `file://` sizing against a real temp file, a `headObject` failure, a range url answered without a request, a range rejected on `file://`, `ContentLength` absent vs `0`, and that every failure arrives through the rejection channel.

`tests/unit/createReadStream.range.test.js` pins the range arithmetic — `?offset=2706&length=332` must ask for `bytes=2706-3037`, `length=1` for `bytes=0-0`, and no range means no `Range` header.

`npm run test:unit` — **19 passing** (4 existing + 15 new).

## Notes for reviewers

- Behaviour change to be aware of: ranged reads now return `length` bytes instead of `length + 1`. For the JSON payloads this package is used for the parsed result is identical, since the extra byte was the newline separator — verified on production data. `length=0` would now produce an invalid range; nothing produces it today, and a loud S3 error beats silently returning one byte.
- Version bumped to **0.7.0** in a separate commit, matching how 0.6.0 was released.
- **Consumer, stated accurately:** cn-core#80 as pushed (`745b047`) still *adds* its own `src/utils/getS3ObjectSize.ts`, still declares `aws-sdk` directly, still pins `v0.4.0`, and still binds `apin.getS3ObjectSize`. The rewrite that deletes that copy, drops the dependency, renames the binding to `apin.getObjectSize` and bumps to `v0.7.0` is prepared but not pushed yet — it lands after this PR merges and is tagged, so the two PRs stop contradicting each other. The name `getObjectSize` wins on both sides, since it sizes `file://` too.

[AP20-813]: https://projectcx.atlassian.net/browse/AP20-813?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by hildebert — APPROVED** [review]

File: src/index.js [hunk 1]
```diff
@@ -46,6 +46,51 @@ exports.createReadStream = function createReadStream(url) {
...
+ *
+ * Accepts every location form createReadStream accepts - an s3:// uri, the
+ * https://s3.amazonaws.com/ equivalent, a file:// path, and the { $src } envelope
+ * upstream steps wrap one in - so that anything streamable is also sizeable.
+ * For S3 this is a headObject call: metadata only, no body transferred.
+ *
+ * @param {string|{$src:string}} url	Remote source location in format s3:// or file://
+ * @return {Promise<number>}	Size of the object in bytes
+ */
+exports.getObjectSize = function getObjectSize(url) {
```
Comment (aleksiiy) [inline src/index.js]: **HIGH — the JSDoc promises `Promise<number>`, but 4 of the 5 failure modes throw synchronously.**

Probed at this head (S3 stubbed, printing `SYNC THROW` vs `REJECTED` per input):

```
range url         -> SYNC THROW Error: getObjectSize does not support range urls
bad format        -> SYNC THROW Error: Unexpected url format
undefined url     -> SYNC THROW TypeError: url argument must be a non-empty String
missing file://   -> SYNC THROW Error: ENOENT ... stat '/definitely/not/here.pdf'
no ContentLength  -> REJECTED Error: Object reported no ContentLength
ContentLength: 0  -> RESOLVED 0
.catch() consumer -> UNCAUGHT SYNC THROW leaks past .catch
```

Two concrete consequences for the consumer this PR exists for:

1. A code operation written the idiomatic way — `apin.getObjectSize(src).then(...).catch(handle)` — never reaches `handle` for a range/format/type error or a missing `file://` path. The throw happens while evaluating the argument, before `.catch` is attached, so it surfaces as an uncaught exception inside the sandbox instead of a handled rejection.
2. `tests/getS3ObjectSize.test.ts` in cn-core#80 has three cases shaped `await expect(getS3ObjectSize(...)).rejects.toThrow(...)`. Swapping in `getObjectSize` — the whole point of this PR — fails all three: the exception fires before `expect` is called.

The test file here already shows the split channel: line 74 asserts with `expect(() => ...).to.throw`, line 82 with a promise handler. Both pass, which is why the asymmetry is invisible in the suite.

Please make this `async function getObjectSize(url)`. Every throw (including `statSync`'s ENOENT) then becomes a rejection, the JSDoc becomes true, and the contract matches the `async` sibling in cn-core.

I don't think "`createReadStream` throws synchronously too" carries here: that function returns a stream, so a synchronous throw is consistent with a synchronous API. This one returns a Promise and says so.

File: package.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "apination-stream-utils",
-  "version": "0.6.0",
+  "version": "0.7.0",
```
Comment (aleksiiy) [inline package.json:3]: **HIGH — `aws-sdk` is a devDependency, but the entry point requires it at load time.** (Commenting here because `package.json` is the file that needs the one-line fix; the declaration itself is at line 27.)

`src/index.js:4` does `const aws = require('aws-sdk')` at module scope, and `index.js` is just `module.exports = require('./src')`. devDependencies are not installed for a package consumed as a dependency, so:

```
$ npm i github:apination/node-stream-utils#v0.6.0
--- aws-sdk present in node_modules? --- 0
--- require the package ---
FAILED: MODULE_NOT_FOUND Cannot find module 'aws-sdk'
```

The package is currently importable only because `aws-sdk` gets hoisted into the consumer's top-level `node_modules` transitively — in cn-core via `cn-utils`; cn-core does not declare it on master (`invokeLambda.ts` even carries an `eslint-disable import/no-extraneous-dependencies` for exactly that).

On its own this is pre-existing and README:131 documents it as "development", so I would not block the package for it. What makes it blocking is this PR's own stated plan: "cn-core PR #80 ... drops its local copy plus the `aws-sdk` direct dependency it needed". Do that, and `require('apination-stream-utils')` fails before `getObjectSize` is ever called — every basic operation that streams, not just sizing.

Since `package.json` is already being touched here, please move `aws-sdk` from `devDependencies` to `dependencies` (or declare it in `peerDependencies` if the intent is that consumers keep supplying it — but then the cn-core cleanup described above must not happen). Same edit in the README dependency list.

File: README.md [hunk 3]
```diff
@@ -24,6 +24,26 @@ stream.on('end', () => {
 });
 ```
 
+## getObjectSize()
```
Comment (aleksiiy) [inline README.md:27]: **MEDIUM — the "Notes for reviewers" description of the consumer does not match cn-core#80's current head** (`745b047`). Anchoring here because the exported name is part of that mismatch.

The PR body says #80 "exposes this to code operations as `apin.getObjectSize` and drops its local copy plus the `aws-sdk` direct dependency it needed. That branch pins `v0.4.0` today and will bump straight to `v0.7.0`." Reading that branch's diff:

- `src/utils/getS3ObjectSize.ts` is **added**, not deleted (58 lines, with its own `S3_URI` copy).
- `package.json` **adds** `"aws-sdk": "^2.1693.0"` — it isn't dropped, it's newly declared.
- The pin is still `github:apination/node-stream-utils#v0.4.0`.
- `src/lib/dependencies/index.ts` binds `apin.getS3ObjectSize`, and `tests/sandboxS3Dependencies.test.ts` asserts that exact name — so this is a rename for the consumer, not just a version bump.

So the migration this PR is justified by isn't staged anywhere yet, and the one part of it that is actionable (dropping `aws-sdk`) is unsafe until the comment on `package.json` is addressed. Either land the corresponding changes on #80, or rewrite this section to describe the actual state — otherwise whoever reviews #80 next reads two PRs that contradict each other about which copy survives and what it's called.

Also worth settling now, since both names are already committed somewhere: `getObjectSize` here vs `getS3ObjectSize` there. The name here is the better one (it covers `file://` too), which means #80's sandbox binding and its test both need updating rather than a pass-through alias.

File: src/index.js [hunk 4]
```diff
@@ -46,6 +46,51 @@ exports.createReadStream = function createReadStream(url) {
...
+				if (head.ContentLength === undefined || head.ContentLength === null) {
+					throw new Error('Object reported no ContentLength: ' + url);
+				}
+
+				return head.ContentLength;
+			});
+	}
+	else if (RX_FILE.test(url)) {
+		const m = url.match(RX_FILE);
+		return Promise.resolve(fs.statSync(m[1]).size);
```
Comment (aleksiiy) [inline src/index.js]: **MEDIUM — range locations are not rejected on this branch, contrary to the README and the comment above.**

`RX_FILE = /^file:\/\/(.+)$/` is greedy, so the query string lands inside the captured path:

```
file:// + range   -> SYNC THROW Error: ENOENT: no such file or directory,
                    stat '/var/folders/.../probe-size.bin?offset=0&length=100'
```

README:44 states "Range locations (`?offset=&length=`) are rejected" without qualification, and the comment at lines 67-69 explains the refusal as a property of the helper. For `file://` the caller instead gets an ENOENT naming a path that doesn't exist, with the range suffix glued on — which reads as "the file is missing" rather than "ranges aren't supported".

Either lift the range check above the branch so it applies to both location forms, or narrow the README sentence and the JSDoc to the S3 forms. First option is preferable — it keeps the two branches honest about the same limitation, and it's the reading a caller takes from the docs.

File: tests/unit/getObjectSize.test.js [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+	// Returning 0 would read as "small" to a caller routing by size, which is the
+	// exact mistake this helper exists to prevent.
+	it('rejects when S3 reports no ContentLength rather than reporting zero', () => {
+		headObjectResult = {};
+
+		return utils.getObjectSize('s3://test-bucket/file.pdf').then(
+			() => { throw new Error('expected rejection'); },
+			err => expect(err.message).to.match(/no ContentLength/)
+		);
+	});
```
Comment (aleksiiy) [inline tests/unit/getObjectSize.test.js:112]: **MEDIUM — the load-bearing part of the `ContentLength` guard is untested.**

The guard at `src/index.js:78` is deliberately `=== undefined || === null` rather than a falsy check, and that distinction is the whole point: a genuinely empty object must return `0`, an unknown size must throw. Verified it behaves correctly today:

```
no ContentLength  -> REJECTED Error: Object reported no ContentLength
ContentLength: 0  -> RESOLVED 0
```

But nothing pins the zero half. A later "simplification" to `if (!head.ContentLength) throw ...` passes all 8 tests in this file while turning every empty object into a hard failure. Please add the counterpart:

```js
it('returns 0 for an object that is genuinely empty', () => {
	headObjectResult = { ContentLength: 0 };

	return utils.getObjectSize('s3://test-bucket/empty').then(size => {
		expect(size).to.equal(0);
	});
});
```

Two smaller gaps in the same suite, worth taking while you're here:

- The stub only ever resolves, so the provider-error path is unexercised — a `NotFound`/`AccessDenied` from `headObject` propagates raw and nothing asserts what a caller sees. One case with `promise: () => Promise.reject(err)` covers it.
- A missing `file://` path isn't covered either; it's the one input where the sync-vs-async channel (first comment) is easiest to regress silently.

File: tests/unit/getObjectSize.test.js [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+	};
+}
+
+function restoreS3() {
+	aws.S3 = originalS3;
+}
+
+const utils = require('../../src');
+
+describe('getObjectSize()', () => {
```
Comment (aleksiiy) [inline tests/unit/getObjectSize.test.js:44]: **LOW — these tests are not reachable from `npm test`, so they gate nothing automatically.** Pre-existing repo gap, not introduced here — flagging it as a note rather than a blocker.

`package.json` has `"test": "mocha tests/integration/"` and `"test:unit": "mocha tests/unit/"`, and the repo has no `.github/` directory at all. So the 12 passing (which I reproduced locally — 4 existing + 8 new, as stated) only happens when someone runs `npm run test:unit` by hand; the default `test` script points at the integration suite, which needs real AWS credentials.

The unit test added with 0.6.0 sits in the same position, so this is convention rather than an oversight in this PR. Since the package is now growing a second consumer-facing entry point, it would be a good moment to either fold the unit dir into `test` (`mocha tests/unit/` first, integration behind its own script) or add a minimal workflow running `npm run test:unit` — `runs-on: self-hosted` per the org convention. Happy for that to be a separate PR.

**Review by aleksiiy — CHANGES_REQUESTED** [review]
Blocking on two one-line items, both inline: `getObjectSize` throws synchronously on 4 of its 5 failure modes despite documenting `Promise<number>` (make it `async`), and `aws-sdk` is declared as a devDependency while the entry point requires it — which makes the `aws-sdk` cleanup promised for cn-core#80 unsafe. Clears once those two land plus a test pinning `ContentLength: 0`.

The remaining comments (consumer description vs cn-core#80's actual head, `file://` range handling, test-script reachability) don't block the merge.

Verification notes: read at `7d65b95`, `npm run test:unit` reproduces 12 passing, and the error-channel / dependency findings come from probes whose output is quoted in the relevant comments.

File: src/index.js [hunk 7]
```diff
@@ -46,6 +46,51 @@ exports.createReadStream = function createReadStream(url) {
...
+		const m = url.match(RX_S3);
+
+		// For a range url createReadStream emits only that slice, and it asks S3 for
+		// `bytes=offset-(offset+length)` - one byte more than length. No single number
+		// describes both the whole object and that stream, so refuse rather than guess.
+		if (m[3]) throw new Error('getObjectSize does not support range urls: ' + url);
+
+		const s3 = new aws.S3();
+
+		return s3.headObject({ Bucket: m[1], Key: m[2] }).promise()
```
Comment (aleksiiy) [inline src/index.js]: **LOW — no `debug()` anywhere on this path, although the module already has a logger.**

`src/index.js:5` sets up `debug('apination:streams')`, and `createWriteStream` uses it for the upload lifecycle. `getObjectSize` makes a real network call to S3 and emits nothing, so when a routing decision goes wrong in production there's no way to tell from CloudWatch which bucket/key was sized or what came back — the only observable is the branch the caller took afterwards.

One line before the call and one in the `then` would be enough, e.g.:

```js
debug(`sizing s3://${m[1]}/${m[2]}...`);
```

and the resolved `ContentLength` on the way out. Keep it to `debug` — nothing here belongs in a customer-facing message.

File: src/index.js [hunk 8]
```diff
@@ -46,6 +46,51 @@ exports.createReadStream = function createReadStream(url) {
...
+	if (typeof url === 'object' && url && (OBJECT_SOURCE_KEY in url)) url = url[OBJECT_SOURCE_KEY];
+	if (typeof url !== 'string' || !url.length) throw new TypeError('url argument must be a non-empty String');
+
+	if (RX_S3.test(url)) {
+		const m = url.match(RX_S3);
+
+		// For a range url createReadStream emits only that slice, and it asks S3 for
+		// `bytes=offset-(offset+length)` - one byte more than length. No single number
+		// describes both the whole object and that stream, so refuse rather than guess.
+		if (m[3]) throw new Error('getObjectSize does not support range urls: ' + url);
```
Comment (aleksiiy) [inline src/index.js]: **LOW — the `createReadStream` off-by-one this refusal rests on needs its own ticket.**

Checked the claim and it holds: `createReadStream` builds `Range: bytes=${offset}-${offset + length}`, and an HTTP byte range is inclusive at both ends, so `?offset=0&length=100` fetches 101 bytes. Every range read through this package returns one byte more than asked for.

Leaving it out of this PR is the right call — fixing it changes existing read behaviour and belongs on its own. But please file it rather than leaving it in a comment here and in the PR body only. Two reasons:

1. It's a live data bug for anyone already using range reads, and right now it's recorded nowhere a ticket search would find it.
2. This block is the only place the reasoning is written down. If someone later corrects the range arithmetic without reading this comment, `length` becomes an honest answer and the refusal above turns into an unexplained restriction — the rationale and the bug need to be linked from both sides.

A line in the ticket pointing back at `src/index.js:70` (and vice versa) is enough.

**Review by aleksiiy — COMMENTED** [review]
Two follow-up notes, neither blocking — the request-changes above still stands on its own two items.

File: src/index.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (ruslan-avantis) [reply src/index.js]: Correct, and fixed in 93b8a88 — though not by deferring the stat: `getObjectSize` is now `async`, which puts *every* failure on the rejection channel rather than only this one. `statSync` inside an async function already rejects, so the sync/async split is gone entirely.

Probed across all seven inputs that used to throw synchronously (bad argument, unsupported format, range url, missing `file://` path); all seven now reject.

File: src/index.js (hunk 1 above)
Comment (ruslan-avantis) [reply src/index.js]: Agreed on every point, fixed in bae870a. `exports.getObjectSize = async function getObjectSize(url)`.

Re-ran your probe shape after the change:

```
undefined                              REJECTED  TypeError
""                                     REJECTED  TypeError
42                                     REJECTED  TypeError
"s3://b/k?offset=0&length=100"         REJECTED  Error
"file:///tmp/x?offset=0&length=100"    REJECTED  Error
"https://example.com/x"                REJECTED  Error
"file:///definitely/not/here.pdf"      REJECTED  Error
```

The test suite now asserts through the rejection channel throughout — a helper that calls `getObjectSize(url).then(reject-on-success, assert)` fails at the call site if anything ever throws synchronously again, so this cannot regress silently.

Your point about `createReadStream` not being a precedent is right: it returns a stream, this returns a Promise and says so.

File: package.json (hunk 2 above)
Comment (ruslan-avantis) [reply package.json:3]: You are right that `devDependencies` is wrong — the entry point requires `aws-sdk` at module scope, and the package is importable today only because consumers hoist it transitively.

Fixed in bae870a, but as `peerDependencies` rather than `dependencies`, which is the second option you offered. Reasoning: a plain dependency would add the whole SDK v2 to **every** consumer bundle, while the Lambda runtime already provides it — a size and cold-start regression imposed on every connector to fix a declaration problem. A peer states the same contract, is checkable by npm, and ships nothing. README dependency list updated alongside.

On the interaction with cn-core#80 that made this blocking: the concern does not apply once #80 is read as rewritten. `main` in cn-core does not declare `aws-sdk` either — the `import/no-extraneous-dependencies` suppression you spotted in `invokeLambda.ts` exists precisely because of that. #80 v1 *added* the declaration; the rewrite removes that addition and returns to `main`'s state. So nothing that works today stops working, and with the peer declared the requirement is now explicit instead of implicit.

File: README.md (hunk 3 above)
Comment (ruslan-avantis) [reply README.md:27]: Fair catch, and the description was the thing that was wrong. Rewritten to state the actual position: #80 as pushed (`745b047`) still adds its own `getS3ObjectSize`, still declares `aws-sdk`, still pins `v0.4.0` and still binds `apin.getS3ObjectSize`. The rewrite that deletes that copy, drops the declaration, renames the binding and bumps to `v0.7.0` is prepared but deliberately not pushed until this PR merges and is tagged — pushing it earlier is what would leave two PRs contradicting each other.

Agreed on the name: `getObjectSize` wins because it sizes `file://` too, so #80's binding and its test move to that name rather than getting an alias.

File: src/index.js (hunk 4 above)
Comment (ruslan-avantis) [reply src/index.js]: Confirmed and fixed, though the resolution changed shape once the off-by-one below was fixed in the same PR.

`file://`: your diagnosis is exact — `RX_FILE` is greedy, the suffix landed in the path, and the caller got ENOENT for a file that never existed. That branch now rejects explicitly with "does not support range urls for file:// locations", since `createReadStream` does not honour ranges there either.

`s3://`: with the arithmetic corrected, `length` *is* the size of the stream, so the helper now returns it straight from the url with no request at all. The README sentence was narrowed to describe both behaviours rather than claiming a blanket refusal.

File: tests/unit/getObjectSize.test.js (hunk 5 above)
Comment (ruslan-avantis) [reply tests/unit/getObjectSize.test.js:112]: Added verbatim, plus the two gaps you listed:

- `returns 0 for an object that is genuinely empty` — pins the `undefined || null` guard, so a "simplification" to `if (!head.ContentLength)` now fails a test instead of passing the suite.
- `propagates a headObject failure as a rejection` — stub rejects with an `AccessDenied`-shaped error.
- `rejects rather than throws on a missing file:// path` — which, as you noted, is the input where the sync/async channel regresses most quietly.

Suite is 19 passing now (4 existing + 15 new), including a new `createReadStream.range` file.

File: tests/unit/getObjectSize.test.js (hunk 6 above)
Comment (ruslan-avantis) [reply tests/unit/getObjectSize.test.js:44]: Leaving this out, and taking your offer of a separate PR — agreed it is worth doing.

Reasoning for not folding it in here: `test` is what "run the tests" means for the whole repository, including anyone's local habit and the `release-*` scripts around it. Repointing it away from the integration suite is a change in the repo's verification contract with its own blast radius, and it would ride along inside a feature PR where nobody is looking for it. Better as its own change where the decision is the subject rather than a side effect.

Worth noting the gap is real either way: the 19 passing here only happen when someone runs `npm run test:unit` by hand.

File: src/index.js (hunk 7 above)
Comment (ruslan-avantis) [reply src/index.js]: Added in bae870a — one line before the call naming bucket and key, one after with the resolved `ContentLength`. Agreed on keeping it to `debug`; a routing decision driven by a silent network call is exactly the thing that is unpleasant to diagnose from CloudWatch after the fact.

File: src/index.js (hunk 8 above)
Comment (ruslan-avantis) [reply src/index.js]: Fixed here instead of filed — 93b8a88.

Checked it against production data before touching it, because the risk was that some producer computed `length` to suit the inclusive range. It does not: at `?offset=2706&length=332` the 332 bytes are exactly one JSON record ending `"COMPANY"}]`, and byte 333 is the `\n` separating the next one. So `length` means what it says, `offset + length - 1` is the correct end, and the bug has been masked all along by the newline separator plus `JSON.parse` tolerating trailing whitespace. Any range read whose next byte is not whitespace has been getting corrupt data.

Re-verified after the fix on the same live object: `loadJson` returns the single expected record, `getObjectSize` on the range returns 332 without a request, and 26,531 on the whole object. `tests/unit/createReadStream.range.test.js` pins the header so this cannot come back quietly.

Your second reason for wanting it written down is what decided the shape of the rest: once the arithmetic is right, the refusal above loses its rationale, so it is gone too and the range branch answers from the url. Fixing both in one diff avoids leaving behind a restriction whose explanation had just been deleted.

**Comment by ruslan-avantis** [comment]
Thanks — the two blocking items are fixed, plus every non-blocking one except the `npm test` wiring, which I have left as the separate PR you suggested. Replies are in each thread; the short version:

| Comment | Resolution |
| --- | --- |
| HIGH — sync throws vs `Promise<number>` | `async` now; all seven previously-throwing inputs reject |
| HIGH — `aws-sdk` devDependency | moved to `peerDependencies`, not `dependencies` — reasoning in thread |
| MEDIUM — `ContentLength: 0` untested | added, plus `headObject` rejection and missing `file://` |
| MEDIUM — `file://` range → ENOENT | rejects explicitly; `s3://` ranges now answer from the url |
| MEDIUM — body vs #80's real head | description rewritten to the actual state |
| LOW — no `debug()` | added around the `headObject` call |
| LOW — off-by-one needs a ticket | **fixed in this PR instead** |
| LOW — `npm test` unreachable | not here; separate PR, per your note |

Two places where I did not take the first option you offered, both explained in thread: `peerDependencies` instead of `dependencies` (a real dependency would add all of aws-sdk v2 to every consumer bundle when the runtime already supplies it), and the range branch of `getObjectSize` now returning `length` instead of refusing (once the arithmetic is correct, the refusal has no rationale left).

The off-by-one turned out to be a live data bug rather than a latent one, and worth reading before merge: `length` means bytes, the code asked for `offset + length`, and it has only ever looked harmless because these payloads are newline-separated JSON. Details and the production verification are in that thread.

19 passing. Ready for another look.

**Review by aleksiiy — APPROVED** [review]
Approving at `93b8a88`. Re-verified against the head rather than the replies: `getObjectSize` is `async` and all six previously-throwing inputs now reject, the `file://` range suffix is refused explicitly instead of surfacing as ENOENT, the three requested tests are present (`npm run test:unit` → 19 passing, reproduced locally), `debug` covers the sizing call, and the consumer section now describes cn-core#80 at `745b047` accurately.

On the off-by-one, which grew out of a request to file a ticket: the fix holds independently of the single payload cited. The producer is `splitStreamIndexedEvent` in pl-workflow-runner (`?offset=' + range[0] + '&length=' + range[1]` from `event.stats.indexes`), and `range[1]` can only be a byte count — at `offset=2706` an end offset of `332` would sit before the start. So `offset + length - 1` is correct, and the extra byte was landing in every ranged read. Worth knowing where that actually mattered: the JSON consumers (`download-from-s3.js` in pl-api-apination and pl-wf-logger, foreach inner steps) tolerated it via `JSON.parse`, but `cn-core/src/lib/utils/http.ts` streams the same value into a multipart body — there it was corrupting the upload, not being ignored.

Two follow-ups, neither blocking the merge:

1. **The README rationale for `peerDependencies` is wrong in both halves and is now written into the repo.** npm 7+ auto-installs peers — a bare consumer with only the git pin gets `aws-sdk@2.1693.0`, 100 MB in `node_modules`, so the tree is identical to what `dependencies` would produce and nothing is saved. And cn-core deploys `nodejs20.x`, whose runtime ships only SDK v3 — v2 was removed after `nodejs16.x`, so it is not "already provided"; it works because `cn-utils` pulls v2 in transitively and it gets zipped. The peer declaration itself is fine (it avoids a second copy when a consumer pins its own version) — it's the two sentences that need correcting, before someone acts on them by adding `peerDependenciesMeta.optional` and reintroducing the `MODULE_NOT_FOUND`.

2. **The range branch returns a number it never verifies.** `s3://b/k?offset=999999&length=100` resolves `100` with no request, while `createReadStream` asks `bytes=999999-1000098` and S3 answers 416 — the stream yields nothing. A range near EOF is clamped the same way: `length` returned, fewer bytes streamed. Safe on the platform path, since engine ranges come from `stats.indexes` computed over the object itself, but a cn-core code operation can build any url from `${...}`. For the signed-upload case this number becomes a `Content-Length`, and a short body stalls the request. One JSDoc sentence stating the range answer is taken from the url and not checked against the object would cover it; `Math.min(length, ContentLength - offset)` costs the same single `headObject` if you'd rather it be verified.

Minor: the peer floor `^2.3.14` isn't exercised by anything (devDep is `^2.814.0`), and the range branch returns before the `debug` line, so the one answer that came from no request also leaves no log.

Don't forget the `v0.7.0` tag — cn-core jumps from `v0.4.0` and picks up the ranged-read change along with the new helper.

File: src/index.js [hunk 9]
```diff
@@ -46,6 +53,68 @@ exports.createReadStream = function createReadStream(url) {
...
+	else if (RX_FILE.test(url)) {
+		// RX_FILE is greedy, so a range suffix ends up inside the path. createReadStream
+		// does not honour ranges for file:// either, so refusing is honest - otherwise
+		// the caller gets ENOENT naming a file that never existed.
+		if (RX_RANGE.test(url)) {
+			throw new Error('getObjectSize does not support range urls for file:// locations: ' + url);
+		}
+
+		const m = url.match(RX_FILE);
+		return fs.statSync(m[1]).size;
```
Comment (apinationbot) [inline src/index.js:111]: **MEDIUM · confirmed** — Local file sizing uses `fs.statSync(m[1]).size` inside an async function. The async wrapper correctly turns a synchronous `ENOENT` into a rejection, but the call still blocks the event loop. For consistency with the S3 `headObject` path and to avoid stalling concurrent sizing of many files, consider using `fs.promises.stat(m[1]).then(s => s.size)` instead.

File: src/index.js [hunk 10]
```diff
@@ -30,7 +31,13 @@ exports.createReadStream = function createReadStream(url) {
...
 			Bucket: m[1],
 			Key: m[2],
-			Range: offset !== null && length !== null ? `bytes=${offset}-${offset + length}` : undefined
+			// An HTTP byte range is inclusive at both ends, so the last byte of a
+			// `length`-byte slice is at `offset + length - 1`. Asking for
+			// `offset + length` returned one byte too many on every range read. It went
+			// unnoticed because the writers of these ranges separate records with a
+			// newline and JSON.parse ignores trailing whitespace - it would corrupt any
+			// range whose next byte is not whitespace.
+			Range: offset !== null && length !== null ? `bytes=${offset}-${offset + length - 1}` : undefined
```
Comment (apinationbot) [inline src/index.js:40]: **POSITIVE · confirmed** — The byte-range arithmetic fix is exactly right: HTTP `bytes=start-end` ranges are inclusive at both ends, so a `length`-byte slice starting at `offset` ends at `offset + length - 1`. The previous `offset + length` requested one byte too many on every ranged read. The new unit tests in `tests/unit/createReadStream.range.test.js` pin the exact `Range` string and the `length=1` edge case.

File: tests/unit/getObjectSize.test.js [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+// Asserts the rejection channel specifically: a synchronous throw from
+// getObjectSize would fail this helper at the call, before .then is attached.
+function expectRejection(url, pattern) {
+	return utils.getObjectSize(url).then(
+		() => { throw new Error('expected a rejection for ' + JSON.stringify(url)); },
+		err => expect(err.message).to.match(pattern)
+	);
+}
+
+describe('getObjectSize()', () => {
```
Comment (apinationbot) [inline tests/unit/getObjectSize.test.js:44]: **POSITIVE · confirmed** — The test suite covers the critical behavior of `getObjectSize` and, importantly, exercises the rejection channel directly: by attaching `.then(...).catch(...)` handlers, a synchronous throw would fail the helper before `.then` runs. This validates the author's claim that every failure mode reaches the caller as a rejection.

**Review by apinationbot — APPROVED** [review]
This PR adds `getObjectSize(url)` alongside the existing stream helpers and fixes the off-by-one byte-range bug in `createReadStream`. It solves the ticket: callers can now size an S3 or `file://` source without downloading it, and the range fix makes the `length` branch of `getObjectSize` internally consistent with what `createReadStream` emits. The unit tests cover the accepted URL forms, `{ $src }` unwrapping, `ContentLength` absent-vs-zero handling, the rejection channel, and the corrected `Range` header arithmetic. `aws-sdk` is correctly declared as a peer dependency, matching the Lambda runtime contract. One maintainability note: `getObjectSize` still uses blocking `fs.statSync` for local paths, but the async wrapper does route the error as a rejection as promised.

**Findings:** 1 MEDIUM, 2 POSITIVE.

**Coverage:** reviewed all 1 slice(s) of this diff.

_Reviewed by George Agent_
